### PR TITLE
DDV all, including domainUnlimited

### PR DESCRIPTION
**Pull request by howard-stearns**
https://highfidelity.manuscript.com/f/cases/21819/Unlimited-Clone-able-items-persist-in-multiple-domains
https://highfidelity.manuscript.com/f/cases/21919/certified-parented-entities-don-t-get-simulated-have-parent-s-fixed-up-or-signals-emitted

### CODE REVIEW

**Comment by howard-stearns** [comment]
Three broad parts:
- Most of this is a **_refactoring_** cleanup:
  - **certificate id => entity id** - A lot of the id being passed around to methods and in packets was a certificate id, and is now an entity id. It didn't matter which was used originally, as they were 1:1, with each certificate id mapping to a single entity id in a domain. But now (below) there may be multiple entity ids for each (domainUnlimited) certificate id.
    - _EntityTree::computeNonce/verifyNonce_ now keep track of the nonce by entity id instead of by cert id. This is mostly just name/type changes to reflect the new behavior.
    - _packet id_ - The various challenge packets and responses, (and the requests for challenges of another interface that are made through the entity server,) all now use an entity id as the id part of the payload, which is used after a full round trip by the original source of the request to match up (possibly multiple overlapping) requests.
      - In some cases, the original sender was mapping this certificate id back to the entity. Now it passes this directly (because there may be multiple entities for each cert). Note that this change does _not_ create a spoofing risk, because the whole challenge consists of verifying the signing of a nonce, and the id (whether cert or entity) has always been used to check that the nonce is, in fact, the one that was sent. So, if someone answered with the wrong entity, the nonce would be wrong, and also the original entity check would go unanswered and fail (by timeout).
      - In other cases, the challenging code didn't pass the entity id around, but side effected the `_lastInspectedEntity` ivar in the ContextOverlayInterface singleton. Now we explicitly pass the entity id around (e.g., `ContextOverlayInterface::startChallengeOwnershipTimer(const EntityItemID& entityItemID)`).
      - While this _is_ a change in packet field size, the id field is specified explicitly in the first few bytes of the packet, so this has always been flexible. The receiver just needs to read the bytes of the given size, and to send them back to the challenger with the required proof. The challenged receiver doesn't have to actually parse the bytes of this id.
      - A lot of the changes in the packet code is not changing anything except the name of this id, to reflect this genericity.
    - _signal arg_ - updateCertificateStatus signal passes an entity id instead of a certificate id as the first arg. (It still isn't _used_ for anything!)
  - The guts of *EntityServer::startDynamicDomainVerification* have been moved to EntityTree::startDynamicDomainVerificationOnServer
    - EntityTree::getEntityCertificateIDMap is no longer needed, because the EntityTree has everything it needs.
    - This avoids a copy of the map from cert to entity id, but it does change the locking.
    - The order of stuff within the loop is different, to make it easier to add an inner-loop below, without having to repeat the cert status lookup on the metaverse.
  - Part of *EntityTree::postAddEntity* has been moved to addCertifiedEntityOnServer.
  - Part of *EntityTree::processRemovedEntities has been moved to removeCertifiedEntityOnServer.
- _**other bug fixes**_
   - **ocking the entity tree during delete** - There are a couple of deleteEntity calls in EntityTree that weren't writeLocking the tree.
   - **Ticket #21919** - when certified entities were RE-added in a domain in which they already existed (so the older one would be deleted), we were bailing out early, so the new one wasn't being added to the simulation, nor having parents/children fixed up, nor having the right signals emitted. This is in EntityTree::postAddEntity.
- _**map of cert-to-entity id is now a map of cert-to-list of entity id**_ - This is the actual behavior change, which, following the above refactoring, is in just four places in EntityTree:
    - The **QHash** ivar mapping certs to entities is now a `QHash<QString, QList<EntityItemID>>`.
    - **EntityTree::addCertifiedEntityOnServer**
    - **EntityTree::removeCertifiedEntityOnServer**
    - **EntityTree::startDynamicDomainVerificationOnServer**

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18433/HighFidelity-Beta-PR15275-3da66da.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18433/HighFidelity-Beta-Interface-PR15275-3da66da.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18433/HighFidelity-Beta-PR15275-3da66da.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18433/HighFidelity-Beta-Interface-PR15275-3da66da.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18433/HighFidelity-Beta-PR15275-3da66da.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18433/HighFidelity-Quest-Beta-PR15275-3da66da.apk)

File: libraries/entities/src/EntityTree.cpp [hunk 1]
```diff
@@ -1421,11 +1396,123 @@ bool EntityTree::isScriptInWhitelist(const QString& scriptProperty) {
...
+        networkRequest.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
+        QUrl requestURL = NetworkingConstants::METAVERSE_SERVER_URL();
+        requestURL.setPath("/api/v1/commerce/proof_of_purchase_status/location");
+        QJsonObject request;
+        request["certificate_id"] = certificateID;
+        networkRequest.setUrl(requestURL);
+
+        QNetworkReply* networkReply = networkAccessManager.put(networkRequest, QJsonDocument(request).toJson());
+
+        connect(networkReply, &QNetworkReply::finished, this, [this, entityIDs, networkReply, minimumAgeToRemove, &certificateID] {
```
Comment (SimonWalton-HiFi) [inline libraries/entities/src/EntityTree.cpp]: Is `entityIDs` safe to use in this lambda? Isn't it a reference to a value in _entityCertificateIDMap that's no longer locked when the network response arrives?

**Pull request merged**